pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package rangefeed
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// feedBudget is memory budget for RangeFeed that wraps BoundAccount
+// and provides ability to wait for downstream to release budget and
+// to send individual events that exceed budget total budget size.
+// feedBudget doesn't provide any fairness when acquiring as it is only
+// supposed to be used by a single caller.
+// When owning component is destroyed, budget could be closed, in that
+// case all budget allocation is returned immediately and no further
+// allocations are possible.
+// In the typical case processor will get allocations from budget which
+// would be in turn borrowed from underlying account. Once event is
+// processed, allocation would be returned.
+// If budget is exhausted, allocations would block until more is available
+// or until an optional timeout expires.
+// NB: if we are competing with other consumers in the budget pool, then
+// we won't see that memory is released.
+type feedBudget struct {
+	mu struct {
+		syncutil.Mutex
+		// Bound account that provides budget with resource.
+		memBudget *mon.BoundAccount
+		// Whenever we pass oversized message this would go to true to prevent
+		// other attempts to acquire more memory until the event goes through.
+		oversize bool
+		// If true, budget was released and no more allocations could take place.
+		closed bool
+	}
+	// Channel to notify that memory was returned to the budget.
+	replenishC chan interface{}
+
+	metrics *Metrics
+}
+
+func (f *feedBudget) init(budget *mon.BoundAccount, metrics *Metrics) {
```

Comment:
Why not use a normal constructor, e.g. `newFeedBudget()`? Doesn't seem like we're pooling these or anything.